code/modules/vehicles/tank.dm

```diff
@@ -0,0 +1,381 @@
+/*
+tankS BY KMC
+THIS IS LIKE REGULAR CM tank BUT IT'S LESS SHIT
+WHOEVER MADE CM TANKS: YOU ARE A BAD CODER!!!!!
+*/
+
+//generic component stuff
+
+/obj/item/tank_weapon //These are really just proof of concept weapons. You'll probably want to adapt them to use the hardpoint crap that CM has.
+	name = "TGS 4 main tank cannon"
+	desc = "A gun that works about 50% of the time, but at least it's open source! It fires tank shells."
+	icon = 'icons/obj/hardpoint_modules.dmi'
+	icon_state = "glauncher"
+	var/obj/item/ammo_magazine/ammo = new /obj/item/ammo_magazine/tank/ltb_cannon
+	var/obj/vehicle/tank/owner
+	var/list/fire_sounds = list('sound/weapons/tank_cannon_fire1.ogg', 'sound/weapons/tank_cannon_fire2.ogg')
+	var/fire_delay
+	var/cooldown = 60 //6 second weapons cooldown
+	var/lastfired = 0 //When were we last shot?
+
+/obj/item/tank_weapon/minigun
+	name = "TGS 3 XENOCRUSHER tank machine gun"
+	desc = "A much better gun that shits out bullets at ridiculous speeds, don't get in its way!"
+	icon = 'icons/obj/hardpoint_modules.dmi'
+	icon_state = "m56_cupola"
+	ammo = new /obj/item/ammo_magazine/tank/ltaaap_minigun
+	fire_sounds = list('sound/weapons/tank_minigun_loop.ogg')
+	cooldown = 2 //Minimal cooldown
+
+/obj/item/tank_weapon/proc/fire(var/atom/T,var/mob/user)
+	if(!can_fire(T))
+		return FALSE
+	if(ammo.current_rounds <= 0)
+		return FALSE
+	if(world.time < lastfired + cooldown)
+		return FALSE
+	lastfired = world.time
+	var/obj/item/projectile/P = new
+	P.generate_bullet(new ammo.default_ammo)
+	log_combat(user, user, "fired the [src].")
+	log_explosion("[user] fired the [src] at [AREACOORD(loc)].")
+	P.fire_at(T, owner, src, P.ammo.max_range, P.ammo.shell_speed)
+	if(!CONFIG_GET(flag/tank_mouth_noise))
+		playsound(get_turf(owner), pick(fire_sounds), 60, 1)
+	ammo.current_rounds--
+	return TRUE
+
+/obj/item/tank_weapon/proc/can_fire(var/turf/T)
+	if(get_dist(T, src) <= 3 && owner.gunner)
+		to_chat(owner.gunner, "Firing your main gun here could damage the tank!")
```

Comment:
Why are you checking for `owner.gunner`? If it's a sanity check, remove it. We want it to runtime, screech and fail if it tries to fire without a gunner owner so we can fix it.

Reply:
This is still a thing.